llvm/lib/CAS/ActionCaches.cpp

```diff
@@ -0,0 +1,225 @@
+  std::array<uint8_t, Size> Value;
+};
+
+class InMemoryActionCache final : public ActionCache {
+public:
+  InMemoryActionCache(CASDB &CAS);
+
+  Error putImpl(ArrayRef<uint8_t> ActionKey, const ObjectRef &Result) final;
+  Optional<ObjectRef> getImpl(ArrayRef<uint8_t> ActionKey) const final;
+  Error poisonImpl(ArrayRef<uint8_t> Key) final;
+
+private:
+  using DataT = CacheEntry<sizeof(HashType)>;
+  using InMemoryCacheT = ThreadSafeHashMappedTrie<DataT, sizeof(HashType)>;
+
+  InMemoryCacheT Cache;
+};
+
+class OnDiskActionCache final : public ActionCache {
+public:
+  Error putImpl(ArrayRef<uint8_t> ActionKey, const ObjectRef &Result) final;
+  Optional<ObjectRef> getImpl(ArrayRef<uint8_t> ActionKey) const final;
+  Error poisonImpl(ArrayRef<uint8_t> Key) final;
+
+  static Expected<std::unique_ptr<OnDiskActionCache>> create(CASDB &CAS,
+                                                             StringRef Path);
+
+private:
+  static StringRef getHashName() { return "BLAKE3"; }
+  static StringRef getActionCacheTableName() {
+    static const std::string Name =
+        ("llvm.actioncache[" + getHashName() + "->" + getHashName() + "]")
+            .str();
+    return Name;
+  }
+  static StringRef getDataPoolTableName() {
+    static const std::string Name =
+        ("llvm.actioncache.data[" + getHashName() + "]").str();
+    return Name;
+  }
+  static constexpr StringLiteral DataPoolFile = "data";
+  static constexpr StringLiteral ActionCacheFile = "actions";
+  static constexpr StringLiteral FilePrefix = "v1.";
+
+  OnDiskActionCache(CASDB &CAS, StringRef RootPath,
+                    OnDiskDataAllocator DataPool,
+                    OnDiskHashMappedTrie ActionCache);
+
+  std::string Path;
+  OnDiskDataAllocator DataPool;
+  OnDiskHashMappedTrie Cache;
+  using DataT = CacheEntry<sizeof(HashType)>;
+  using ActionCacheResultT = std::atomic<DataT *>;
+};
+} // end namespace
+
+// TODO: Check the hash schema is the same between action cache and CAS. If we
+// can derive that from static type information, that would be even better.
+InMemoryActionCache::InMemoryActionCache(CASDB &CAS) : ActionCache(CAS) {}
+
+OnDiskActionCache::OnDiskActionCache(CASDB &CAS, StringRef Path,
+                                     OnDiskDataAllocator DataPool,
+                                     OnDiskHashMappedTrie Cache)
+    : ActionCache(CAS), Path(Path.str()), DataPool(std::move(DataPool)),
+      Cache(std::move(Cache)) {}
+
+Optional<ObjectRef> InMemoryActionCache::getImpl(ArrayRef<uint8_t> Key) const {
+  auto Result = Cache.find(Key);
+  if (!Result)
+    return None;
+  return getCAS().getReference(Result->Data.getValue());
+}
+
+static Error createResultCachePoisonedError(StringRef Key, StringRef Output,
+                                            StringRef ExistingOutput) {
+  return createStringError(std::make_error_code(std::errc::invalid_argument),
+                           "cache poisoned for '" + Key + "' (new='" + Output +
```

Comment:
This terminology is a bit confusing: we have a `poison` method  that presumably mutates the cache, but we are also returning an error for "poisoning" when the cache has not been mutated (the existing value is left in place).

Also, "poison" is not a great term for this: in the context of the compiler it's more likely this represents non-determinism not a deliberate cache poisoning attack.  I would reword this to something like "cannot modify existing value for key".